test/integration/replicaset/replicaset_test.go

```diff
@@ -250,11 +473,9 @@ func TestAdoption(t *testing.T) {
 			t.Fatalf("Failed to create Pod: %v", err)
 		}
 
-		stopCh := make(chan struct{})
-		informers.Start(stopCh)
-		waitToObservePods(t, podInformer, 1)
-		go rm.Run(5, stopCh)
-		if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
+		stopCh := runControllerAndInformers(t, rm, informers, 1)
+		defer close(stopCh)
```

Comment:
This may not be the root/only cause, but this is one source of short-term goroutine bloat. Previously, `close(stopCh)` was called at the end of each iteration. Now, they are all queued up and called when `TestAdoption` returns. This allows more build-up of goroutines during the test.

There is also a similar problem above, which predates your PR. The `defer closeFn()` being inside the loop is going to result in many more simultaneously running apiservers than intended.

To keep the defer logic clear, you can put the whole loop body in a function:

```go
for i, tc := range testCases {
  func() {
    ...
    defer ...
  }()
}
```

Reply:
Ah. I reverted the codes for running controller and informers to check whether the error still happened. It did, so it is an existing problem predates this PR. I think putting the whole loop body in a function will work. As soon as the function returns, the channel is closed.

Reply:
Can you push a new commit where this is fixed? Although it wasn't the root cause, it's still more correct to clean up after each iteration.